docs/developer/solutions/archive.rst

```diff
@@ -0,0 +1,152 @@
+       -R  # (or "-rock" if available)
+       -J  # (or "-joliet" if available)
+       -joliet-long
+       -l  # (or "-full-iso9660-filenames" if available)
+       -V 'MySolution 1.0.0'  # (or "-volid" if available)
+       -gid 0
+       -uid 0
+       -iso-level 3
+       -input-charset utf-8
+       -output-charset utf-8
+       my_solution_root/
+
+.. _mkisofs: https://linux.die.net/man/8/mkisofs
+
+.. todo::
+
+   Consider if overriding the source files UID/GID to 0 is necessary
+
+
+File hierarchy
+--------------
+
+Here is the file tree expected by MetalK8s to exist in each Solution archive::
+
+   .
+   ├── images
+   │   ├── listing.yaml
+   │   └── some_image_name.tar.gz
+   ├── operator
+   │   ├── crd
+   │   │   └── some_crd_name.yaml
+   │   └── deployment.yaml
+   ├── product.txt
+   └── ui
+       └── deployment.yaml
+
+
+Product information
+-------------------
+
+General product information about the packaged Solution must be stored in the
+``product.txt`` file, stored at the archive root.
+
+It must respect the following format::
+
+   NAME=${YOUR_SOLUTION_NAME}
+   VERSION=${YOUR_PROJECT_FULL_VERSION}
+
+.. todo::
+
+   Consider if more keys are needed, e.g. author/maintainer, Git reference
+
+
+.. _solution-archive-images:
+
+OCI images
+----------
+
+.. todo::
+
+   With `#1049`_, we will have another standard way of serving OCI images with
+   a read-only registry directly exposing the ISO contents.
+   A specific process will be required to package these images, which we need
+   to document here.
+
+.. _`#1049`: https://github.com/scality/metalk8s/issues/1049
+
+MetalK8s exposes container images in the OCI_ format through a registry.
+Solutions should thus provide their container images in a compatible format,
+for example using ``docker save``.
+
+For size concerns, we expect such images to be compressed using ``gzip``.
+
+Here is an example of how to build, save and compress an image with
+``docker``::
+
+   docker build --tag my-image:1.0.0 --file my_image.dockerfile
+   docker save my-image:1.0.0 -o my_image_1_0.tar
+   gzip -9 my_image_1_0.tar
+
+.. _OCI: https://github.com/opencontainers/image-spec/blob/master/spec.md
+
+MetalK8s also requires that container images respect specific standards, such
+as using the same base image or having appropriate ``LABEL`` statements, as
+described in :ref:`req-container-images`.
+
+In order for MetalK8s to populate its registry with accurate image tags,
+Solutions must provide a ``listing.yaml`` file under ``/images``, with the
+given format::
+
+   apiVersion: metalk8s.scality.com/v1alpha1
+   kind: ImagesList
+   solution: MySolution
+   images:
+     - archive: my_image_1_0.tar.gz
+       tag: my_image:1.0.0
+
+.. note::
+
+   Each Solution ISO will be made available to Pods as a single repository.
+   Operator Deployments will then be required to accept the repository prefix,
+   stored in a ``ConfigMap``, in order to properly configure the resources they
```

Comment:
Operators need to know _where_ the images exist, so they can reference them in the various objects they create. What we suggest here is that we will define a `ConfigMap` to mount as an environment variable, for Operators to use as their `REPOSITORY_PREFIX`.

Let's take our "Example" solution described in the comment above:

- I have imported "Example v1.0", images are available in some repository
- I retrieve the full URL of the Operator image, including the repository (_Note: this is cumbersome, we may want to find a better way_)
- I deploy the Operator for this version, passing into the container spec:

```
image: {{ full_operator_image_url }}
[...]
env:
- name: REPOSITORY_PREFIX
    valueFrom:
    configMapKeyRef:
        name: metalk8s-solutions-repositories
        key: example-1.0
```

- The Operator container can now use this to generate its own resources, using `image: {{ REPOSITORY_PREFIX }}/example:1-latest`

Now, suppose (as in the comment above) that I have deployed version 1.1 of my Example solution. If I want to upgrade the 1.0 Operator, I can edit its Deployment, to use the new Operator image, and pass change the ConfigMap key to use, which will effectively upgrade all instances of `example:1-latest` deployed.

I know this all feels complicated, and thanks a lot for taking the time to review that part. We'll clarify these processes as soon as we know more (working on a _Hello-world Solution_ in the coming weeks should help us on that).

Reply:
Once each operator knows where to get images it's pretty straightforward and it's good that it's "managed" by the deployer that way.

> I retrieve the full URL of the Operator image, including the repository (Note: this is cumbersome, we may want to find a better way)

For bootstrapping each Operator and making the initial manifests configurable, any opinions on `kustomize` / `kubectl apply -k`? It can be used [as an API](https://github.com/kubernetes-sigs/kustomize/blob/master/pkg/types/kustomization.go#L42) and allows replacing images (in addition to general patching) pretty easily as input data, leaving operator deployment manifests "clean". I can see how it only shifts one part of the problem elsewhere and does not entirely solve it though :)

Reply:
I'll look into `kustomize`, thanks for the suggestion :)

Reply:
So, I looked into `kustomize`, it's really interesting. It doesn't seem to be available as part of `apiserver`, only inside `kubectl`, but it could be a good addition to the manifests provided in the ISO. Let's wait until we know more with #1060, I'll try it out.

I added a note to mention this in the docs.